src/transformers/trainer.py

```diff
@@ -2382,7 +2382,7 @@ def _save_checkpoint(self, model, trial, metrics=None):
             self._push_from_checkpoint(staging_output_dir)
 
         # Place checkpoint in final location after all saving is finished.
-        if staging_output_dir != output_dir:
+        if staging_output_dir != output_dir and self.args.should_save:
```

Comment:
Is there a race condition here, where the main process renames the staging directory before other writing processes are finished checkpointing?

Reply:
Yes indeed, we can use `main_process_first` here and check if it exists before renaming. This will ensure that everything will be renamed and process execution must happen in order. 

Specifically it will make sure that the rename happens first before anything else is allowed to continue

Reply:
Do you want 'main process first' or 'main process last'? I'd have thought you'd want to delay the rename until all other processes have stopped writing. This looks like the rename will happen too early. 

The main process could maybe symlink the staging directory to the final destination directory.

Reply:
Even if main process is last, that wouldn't really do anything. I can do an alternative though.

Reply:
The good alternative is `wait_for_everyone` which will, as suggested, wait for all processes to reach that point before proceeding. (Just pushed)